neo4j/internal/bolt/connect.go

```diff
@@ -115,3 +134,181 @@ func Connect(ctx context.Context,
+			errorListener.OnDialError(ctx, serverName, err)
+		}
+		return 0, 0, err
+	}
+
+	// Send the handshake confirmation.
+	if err = sendHandshakeConfirmation(ctx, conn, boltLogger, errorListener, serverName, chosen, capBytes); err != nil {
+		return 0, 0, err
+	}
+
+	return chosen.major, chosen.minor, nil
+}
+
+// readProtocolOfferings reads the number of protocol offerings and returns the count and
+// a slice of supported protocol versions.
+func readProtocolOfferings(ctx context.Context, r racing.RacingReader, serverName string, errorListener ConnectionErrorListener) (uint64, []protocolVersion, error) {
+	count, err := readVarInt(ctx, r)
+	if err != nil {
+		errorListener.OnDialError(ctx, serverName, err)
+		return 0, nil, fmt.Errorf("failed to read manifest protocol count: %w", err)
+	}
+	supported := make([]protocolVersion, count)
+	for i := uint64(0); i < count; i++ {
+		var versionBytes [4]byte
+		_, err := r.ReadFull(ctx, versionBytes[:])
+		if err != nil {
+			errorListener.OnDialError(ctx, serverName, err)
+			return 0, nil, fmt.Errorf("failed to read manifest protocol version: %w", err)
+		}
+		supported[i] = protocolVersion{
+			back:  versionBytes[1],
+			minor: versionBytes[2],
+			major: versionBytes[3],
+		}
+	}
+	return count, supported, nil
+}
+
+// readCapabilityMask reads the capability bit mask (a Base128 VarInt) and returns both the
+// raw value and its encoded byte slice.
+func readCapabilityMask(ctx context.Context, r racing.RacingReader, serverName string, errorListener ConnectionErrorListener) (uint64, []byte, error) {
+	capMask, err := readVarInt(ctx, r)
+	if err != nil {
+		errorListener.OnDialError(ctx, serverName, err)
+		return 0, nil, fmt.Errorf("failed to read capability mask: %w", err)
+	}
+	capBytes, err := encodeVarInt(capMask)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to encode capability mask: %w", err)
+	}
+	return capMask, capBytes, nil
+}
+
+// logManifestHandshake logs the complete server handshake message for manifest negotiation.
+// It prints the initial response, count of offerings, each supported protocol, and the capability mask.
+func logManifestHandshake(boltLogger log.BoltLogger, response []byte, count uint64, supported []protocolVersion, capBytes []byte) {
+	if boltLogger == nil {
+		return
+	}
+	var supportedProtocols []string
+	for _, p := range supported {
+		supportedProtocols = append(supportedProtocols, p.formatProtocol())
+	}
+	boltLogger.LogServerMessage("", "<HANDSHAKE> %s [%d] %s %s",
+		fmt.Sprintf("%#X", response),
+		count,
+		strings.Join(supportedProtocols, " "),
+		fmt.Sprintf("%#X", capBytes))
+}
+
+// selectProtocol iterates over our protocol proposals (skipping the manifest marker)
+// and returns the first candidate whose major version matches and whose minor version
+// falls within the range offered by the server.
+func selectProtocol(supported []protocolVersion, errorListener ConnectionErrorListener, serverName string) (protocolVersion, error) {
```

Comment:
```suggestion
func selectProtocol(supported []protocolVersion) (protocolVersion, error) {
```
Unused parameters. However, maybe `errorListener` should actually be used instead of being removed.